crates/jit/src/code_memory.rs

```diff
@@ -2,14 +2,45 @@
 
 use crate::function_table::FunctionTable;
 use region;
+use std::mem::ManuallyDrop;
 use std::{cmp, mem};
 use wasmtime_environ::{Compilation, CompiledFunction};
 use wasmtime_runtime::{Mmap, VMFunctionBody};
 
+struct CodeMemoryEntry {
+    mmap: ManuallyDrop<Mmap>,
+    table: ManuallyDrop<FunctionTable>,
+}
+
+impl CodeMemoryEntry {
+    fn new() -> Self {
+        Self {
+            mmap: ManuallyDrop::new(Mmap::new()),
+            table: ManuallyDrop::new(FunctionTable::new()),
+        }
+    }
+    fn with_capacity(cap: usize) -> Result<Self, String> {
+        Ok(Self {
+            mmap: ManuallyDrop::new(Mmap::with_at_least(cap)?),
+            table: ManuallyDrop::new(FunctionTable::new()),
+        })
+    }
+}
+
+impl Drop for CodeMemoryEntry {
+    fn drop(&mut self) {
+        unsafe {
+            // Table needs to be freed before mmap.
+            ManuallyDrop::drop(&mut self.table);
+            ManuallyDrop::drop(&mut self.mmap);
+        }
+    }
+}
+
 /// Memory manager for executable code.
 pub struct CodeMemory {
-    current: (Mmap, FunctionTable),
-    mmaps: Vec<(Mmap, FunctionTable)>,
+    current: CodeMemoryEntry,
+    mmaps: Vec<CodeMemoryEntry>,
```

Comment:
Extreme nit (feel free to ignore):

```suggestion
    entries: Vec<CodeMemoryEntry>,
```
```